source/uri-options/tests/read-preference-options.json

```diff
@@ -0,0 +1,65 @@
+{
+  "tests": [
+    {
+      "description": "Valid connection and timeout options are parsed correctly",
+      "uri": "mongodb://example.com/?readPreference=primaryPreferred&readPreferenceTags=dc:ny,rack:1&maxStalenessSeconds=120&readPreferenceTags=dc:ny",
+      "valid": true,
+      "warning": false,
+      "hosts": null,
+      "auth": {
+        "username": null,
+        "password": null,
+        "db": null
+      },
+      "options": {
+        "readPreference": "primaryPreferred",
+        "readPreferenceTags": [
+          "dc:ny,rack",
+          "dc:ny"
+        ],
+        "maxStalenessSeconds": "120"
+      }
+    },
+    {
+      "description": "Arbitrary string read preference does not cause a warning",
+      "uri": "mongodb://example.com/?readPreference=arbitraryButStillValid",
```

Comment:
Can you clarify why this does not raise a warning? It seems odd that we do validate the `readPreferenceTags` (since an invalid tag raises a warning) but don't do the same for the `readPreference` itself.

Reply:
This was more of a judgment call than anything else. In other places (like read/write concerns), we explicitly don't throw errors on unknown levels to make things more forward compatible. Read preferences are a bit different given that they're only sent to mongos's (and not mongod's), so some level of validation will need to be done in the driver, but I wasn't sure if this was something that should be done in the connection string parsing or not. The distinction between this and the tags is that the warning for tags is about the _format_ of the value, not the value itself. While we might someday add an additional string that is a valid read preference, we aren't likely to allow tag sets that are anything besides `string -> string` maps.

That being said, given that this is a judgment call, maybe the best solution is to just remove this test entirely and not mandate either a warning or a lack thereof. Assuming that's acceptable to you, I can push this change tomorrow morning.

Reply:
I had thought about the point you make about the warning in tags stemming from the formatting of content and not the content itself. I think that's a reasonable rule to live by as far as this spec is concerned. 

To avoid confusion, can we change the 'Accepted values' for `readPreference` to say something like 

> any string; currently supported read preferences are 'primary', 'primaryPreferred', ...

The current value does seem to imply that the `readPreference` value _needs_ to be one of the strings mentioned there or it will be invalid. Conversely, we can also simply point the reader to the read preference spec to see for themselves what read preferences are supported.

Reply:
> Conversely, we can also simply point the reader to the read preference spec to see for themselves what read preferences are supported.

Actually, this doesn't seem possible since the spec makes no mention of the camel case names of read preferences: https://github.com/mongodb/specifications/blob/master/source/driver-read-preferences.rst#mode

Reply:
Changing it to a link sounds good to me; I'll update the doc accordingly

Reply:
Hmm, good catch. It does seem to mention them below in the "communication with mongos" section. I'll update the language to reflect that camelCase is expected.

Reply:
Sorry for being a bit late to the thread here, but I would argue that drivers should enforce unknown read preference modes. I don't think this prior comment from @saghm was entirely accurate:

> Read preferences are a bit different given that they're only sent to mongos's (and not mongod's), so some level of validation will need to be done in the driver, but I wasn't sure if this was something that should be done in the connection string parsing or not.

Drivers do use read preferences locally to perform [server selection](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-preference), and that spec outlines the five supported modes.

For prior art, I can tell you that both libmongoc and PHP report errors for unsupported modes:

 * https://github.com/mongodb/mongo-c-driver/blob/1.13.0/src/libmongoc/src/mongoc/mongoc-uri.c#L1318
 * https://github.com/mongodb/mongo-php-driver/blob/1.5.3/src/MongoDB/ReadPreference.c#L80

IMO, validating that the mode string is one of five possible values (after canonicalization; note that libmongoc and PHP use `strcasecmp()`) is sensible to avoid errors down the line.

Reply:
> I don't think this prior comment from @saghm was entirely accurate

I think we're saying the same thing? I was saying that since we don't just forward read preferences to all server types (like we do with read/write concerns) and actually use them internally, the driver will need to validate them. I'm not arguing that drivers shouldn't validate them, which would be impossible; I'm just saying that I'm not convinced that the spec should enforce that the drivers have to validate them _here_.

Reply:
Sounds good. I likely focused too much on the "only sent to mongos" bit of your last comment and didn't fully process the "validation will need to be done in the driver" bit.

If this spec does not intend to dictate read preference validation, would it be preferable to simply omit any tests for invalid mode strings? As-is, the "Arbitrary string read preference does not cause a warning" test may be requiring a driver to _not_ validate read preferences during URI parsing.

Reply:
Yes, that's what we decided; check the current diff, the test is already gone